jolt-core/src/jolt/vm/registers/read_write_checking.rs

```diff
@@ -436,205 +448,358 @@ impl<F: JoltField> RegistersReadWriteChecking<F> {
-                            rs2_ra[0][k] += A[j_bound];
 
-                            let k = trace[j].rd_write().0;
-                            unsafe {
-                                dirty_indices.insert_unchecked(k);
+                            for k in dirty_indices.ones() {
+                                val_j_r[0][k] = val_j_0[k];
+                            }
+                            let mut inc_iter = inc_chunk.iter().peekable();
+
+                            // First of the two rows
+                            loop {
+                                let (row, col, inc_lt, inc) = inc_iter.next().unwrap();
+                                debug_assert_eq!(*row, j_prime);
+                                val_j_r[0][*col] += *inc_lt;
+                                val_j_0[*col] += *inc;
+                                if inc_iter.peek().unwrap().0 != j_prime {
+                                    break;
+                                }
+                            }
+                            for k in dirty_indices.ones() {
+                                val_j_r[1][k] = val_j_0[k];
                             }
-                            rd_wa[0][k] += A[j_bound];
-                        }
 
-                        for j in (j_prime + 1) << round..(j_prime + 2) << round {
-                            let j_bound = j % (1 << round);
+                            // Second of the two rows
+                            for inc in inc_iter {
+                                let (row, col, inc_lt, inc) = *inc;
+                                debug_assert_eq!(row, j_prime + 1);
+                                val_j_r[1][col] += inc_lt;
+                                val_j_0[col] += inc;
+                            }
 
-                            let k = trace[j].rs1_read().0;
-                            unsafe {
-                                dirty_indices.insert_unchecked(k);
+                            let eq_r_prime_eval = gruens_eq_r_prime.E_out_current()[j_prime / 2];
+                            let inc_cycle_evals = {
+                                let inc_cycle_0 = inc_cycle.get_bound_coeff(j_prime);
+                                let inc_cycle_1 = inc_cycle.get_bound_coeff(j_prime + 1);
+                                let inc_cycle_infty = inc_cycle_1 - inc_cycle_0;
+                                [inc_cycle_0, inc_cycle_infty]
+                            };
+
+                            let mut inner_sum_evals = [F::zero(); DEGREE - 1];
+                            for k in dirty_indices.ones() {
+                                if !rs1_ra[0][k].is_zero()
+                                    || !rs1_ra[1][k].is_zero()
+                                    || !rs2_ra[0][k].is_zero()
+                                    || !rs2_ra[1][k].is_zero()
+                                    || !rd_wa[0][k].is_zero()
+                                    || !rd_wa[1][k].is_zero()
```

Comment:
the reason I had multiple `if` blocks before was as an optimization; let's go back to doing that bc I think that's why we're seeing a speed up for RAM read-write checking but not for registers